src/lib/Zikula/Core/Theme/Engine.php

```diff
@@ -0,0 +1,243 @@
+    {
+        $request = $requestStack->getCurrentRequest();
+        if (!empty($request)) {
+            $themeName = $this->getCurrentTheme($request);
+            // @todo remove usage of ThemeUtil class
+            $this->themeBundle = \ThemeUtil::getTheme($themeName);
+            $this->setRequestAttributes($request);
+        }
+    }
+
+    /**
+     * @api
+     * Initialize the theme engine based on the Request
+     * @param Request $request
+     */
+    public function setRequestAttributes(Request $request)
+    {
+        $this->requestAttributes = $request->attributes->all();
+        $this->requestAttributes['pathInfo'] = $request->getPathInfo();
+    }
+
+    /**
+     * @api
+     * wrap the response in the theme.
+     *
+     * @param Response $response @todo change typecast to ThemedResponse in 2.0
+     * @return Response|bool (false if theme is not twigBased)
+     */
+    public function wrapResponseInTheme(Response $response)
+    {
+        $this->overrideThemeIfRequired($response);
+
+        // original OR overridden theme may not be twig based
+        // @todo remove twigBased check in 2.0
+        if (!$this->themeBundle->isTwigBased()) {
+            return false;
+        }
+
+        return $this->themeBundle->generateThemedResponse($response);
+    }
+
+    /**
+     * @api
+     * wrap a block in the theme's block template
+     * @todo consider changing block to a Response
+     *
+     * @param array $block
+     * @return bool|string (false if theme is not twigBased)
+     */
+    public function wrapBlockInTheme(array $block)
+    {
+        // @todo remove twigBased check in 2.0
+        if (!$this->themeBundle->isTwigBased()) {
+            return false;
+        }
+
+        return $this->themeBundle->generateThemedBlock($block);
+    }
+
+    /**
+     * @deprecated This will not be needed >=2.0 (when Smarty is removed)
+     * may consider leaving this present and public in 2.0 (unsure)
+     * @return string
+     */
+    public function getThemeName()
+    {
+        return $this->themeBundle->getName();
+    }
+
+    /**
+     * @deprecated This will not be needed >=2.0 (when Smarty is removed)
+     * @return bool
+     */
+    public function themeIsOverridden()
+    {
+        return $this->themeIsOverridden;
+    }
+
+    /**
+     * @api
+     * @param Request $request
+     * @return \Zikula\Core\AbstractTheme
+     */
+    public function getTheme(Request $request = null)
+    {
+        return $this->themeBundle;
+    }
+
+    /**
+     * Find the realm in the theme.yml that matches the given path, route or module
+     * Uses regex to match a pattern to one of three possible values
+     *
+     * @todo is there a faster way to do this?
+     * @return int|string
+     */
+    private function setMatchingRealm()
+    {
+        foreach ($this->themeBundle->getConfig() as $realm => $config) {
+            if (!empty($config['pattern'])) {
```

Comment:
I don't like how the pattern is matched against everything - I'd like to see three distinct settings for these.